src/App.tsx

```diff
@@ -1,17 +1,126 @@
-import React from "react";
-import { BrowserRouter as Router, Route } from "react-router-dom";
+import React, { useEffect, useState, Component } from "react";
+import {
+  BrowserRouter as Router,
+  Route,
+  RouteProps,
+  Redirect
+} from "react-router-dom";
+
 import "./App.css";
-import HomePage from "./pages/HomePage";
+import ReelSpinner from "./components/ReelSpinner";
+import { Paths } from "./constants";
+import ExplorePage from "./pages/ExplorePage";
+import LandingPage from "./pages/LandingPage";
 import LoginPage from "./pages/LoginPage";
+import MyMoviesPage from "./pages/MyMoviesPage";
 import SignupPage from "./pages/SignupPage";
+import { User } from "./types";
+
+interface AuthenticationRouteProps extends RouteProps {
+  component: React.ComponentType<any>;
+  sessionUser: User | null;
+  setSessionUser: (user: User | null) => void;
+}
+
+const AuthenticatedRoute = ({
+  component: Component,
+  sessionUser,
+  setSessionUser,
+  ...rest
+}: AuthenticationRouteProps) => (
+  <Route
+    {...rest}
+    render={() =>
+      Boolean(sessionUser) ? (
+        <Component sessionUser={sessionUser} setSessionUser={setSessionUser} />
```

Comment:
Ended up going w/ `children` (instead of keeping a prop) because it looks a lot nicer and is easier to reason about.

It's not perfect, b/c typescript doesn't know that `sessionUser` cannot be null at that stage, so you need to use a `!`.